src/signature_builder.rs

```diff
@@ -0,0 +1,145 @@
+//! signature index construction as builder pattern
+
+use super::*;
+
+use std::default::Default;
+
+// marker trait for stages
+pub trait ConstructionStage {}
+
+pub struct Empty;
+pub struct WithDigest;
+pub struct WithSignature;
+
+impl ConstructionStage for Empty {}
+
+impl ConstructionStage for WithDigest {}
+
+impl ConstructionStage for WithSignature {}
+
+/// base signature header builder
+///
+/// T describes the stage and can be one of Empty, WithDigest, WithSignature
+pub struct SignatureHeaderBuilder<T>
+where
+    T: ConstructionStage,
+{
+    entries: Vec<IndexEntry<IndexSignatureTag>>,
+    // TODO right now this is always 0
+    // TODO verify what this actually is
+    // TODO and how it must be constructed
+    offset: i32,
```

Comment:
Setting it to 0 is fine since the offset will be calculated for you in `from_entries`. I think you can remove this property.